Fix UDP/ICMPv6 checksum for a sliced/accumulated payload.

kb2ma opened the pull request:
The checksum calculation for UDP and ICMPv6 allows for looping through multiple payload snips. However, inet_csum() always zero-padded on an odd number of bytes, even if not the last snip in the payload.

I have not yet looked into unit testing in RIOT, but I plan to create a test case for this situation. First, I wanted to confirm my solution is reasonable.

You might ask why someone would need multiple payload snips. Well, a [GNRC CoAP implementation](https://github.com/kb2ma/RIOT/tree/gcoap/client) might use such a thing. :-) First things first.

kb2ma commented:
Looks like I've got a problem with whitespace. I will look into it.

cgundogan commented:
Nice! I didn't look at your proposed change yet, but having this issue addressed is a significant step forward. Having a wrongly calculated checksum on an odd number of bytes is a known bug, although I coudn't find a corresponding issue on github.

authmillenon commented:
Fixes #3201.

authmillenon commented:
Thanks for providing a fix for that. Can you please provide a unittest for this situation in https://github.com/RIOT-OS/RIOT/blob/master/tests/unittests/tests-inet_csum/tests-inet_csum.c.

authmillenon commented on sys/include/net/inet_csum.h:
```diff
@@ -35,14 +37,41 @@ extern "C" {
  * @return  The unnormalized Internet Checksum of @p buf.
  */
-uint16_t inet_csum(uint16_t sum, const uint8_t *buf, uint16_t len);
+uint16_t inet_csum_slice(uint16_t sum, const uint8_t *buf, uint16_t len, size_t accum_len);
```
This does not need to be public IMHO

authmillenon replied:
Nvm, did not see how you use it.

kb2ma commented:
Thanks for the quick review and pointer to the unit test code. I'll dig in to that and get back soon.

authmillenon commented:
One minor thing (that's just a git-usage style remark so completely optional): Would you mind dividing the API change to `net/inet_csum.h` and the usage of the new functions in the layer 4 modules into at least two separate commits to keep the changes independent module-wise?

kb2ma commented:
I have updated the PR with commits for the unit test, whitespace fix, and separation of layer 4 files. If it looks good to @authmillenon, I plan to squash into two commits:

1. From the original commit, 3695430, through 8b94da0, which removed the L4 files.
2. The final commit, 87355c9, which re-added the L4 files.

I see Travis is complaining, but it looks like that is only due to the need to squash. Please let me know if I have missed something.

authmillenon commented:
ACK, if travis is happy. I would suggest however to make it three commits:

1. API change
2. tests
3. L4 files

authmillenon commented:
(please squash)

kb2ma commented:
@authmillenon -- I have a git push question. I have squashed the commits on my workstation down to the three you suggested. However, when I try to push to GitHub, I see the error below. Do I want to use 'push -f'? Below the error, I list the squashed commits. (FWIW, I reworded the original commit.) I understand that I have essentially rewritten this branch with the squash. But before I get rid of the remote history, I want to be sure the squash will work with this pull request. Thanks!

> kbee@gazelle:~/dev/riot/repo$ git push origin misc/fix-checksum 
> ! [rejected]        misc/fix-checksum -> misc/fix-checksum (non-fast-forward)
> error: failed to push some refs to 'git@github.com:kb2ma/RIOT.git'
> hint: Updates were rejected because the tip of your current branch is behind
> hint: its remote counterpart. Integrate the remote changes (e.g.
> hint: 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.

Squashed commits below:

commit c9338a4557b073e2f41d079b55bdebce932ffc3c
Author: Ken Bannister <kb2ma@runbox.com>
Date:   Sat Nov 28 16:11:07 2015 -0500

    Update layer 4 files to fix checksum with inet_csum_slice().

commit 58a2bb187fbd9da12cdb55c4f2889ae3593591b3
Author: Ken Bannister <kb2ma@runbox.com>
Date:   Fri Nov 27 23:50:09 2015 -0500

    Add unit test for inet_csum_slice()

commit 4a226cfb215466fe7b0f5c4d9cfdbf5e23809ced
Author: Ken Bannister <kb2ma@runbox.com>
Date:   Wed Nov 25 23:33:28 2015 -0500

    Add inet_csum_slice() to fix checksum for a sliced layer 4 payload
    
    Padding for an odd number of bytes was not calculated properly.

authmillenon commented:
Yes, you have rewritten the history, so you MUST `git push -f`. For a public master or development branch this is usually a no-go, but for feature branches (that's what your PR branch is) this is quite common, to keep the master-history clean. Your history won't be lost, you can still use `git reflog` to find the older commit hashes (which apply to your local repository as well as the remote repository).

kb2ma commented:
Squashed; thanks for the guidance on git push. Travis's only complaint at the moment looks to be the 'needs squashing' label.

Also, I've reviewed the commits closely over the past day. There is one issue that bugs me, and I'd like to know your level of concern. inet_csum_slice() will fail if a zero length buffer is passed to it *and* the accumulated length is an odd number. This scenario doesn't happen in practice because a pktsnip can't have a zero length buffer. However, inet_csum_slice() is written to accept a generic buffer, and not just a pktsnip.

The old implementation of inet_csum() was robust against a zero length buffer. The fix to inet_csum_slice() is simple enough -- to wrap the relevant portion of the function in a conditional on the length parameter. I would be happy to do this in this PR, or in a follow-on improvement PR.

authmillenon commented:
> Also, I've reviewed the commits closely over the past day. There is one issue that bugs me, and I'd like to know your level of concern. inet_csum_slice() will fail if a zero length buffer is passed to it and the accumulated length is an odd number. This scenario doesn't happen in practice because a pktsnip can't have a zero length buffer. However, inet_csum_slice() is written to accept a generic buffer, and not just a pktsnip.

Thanks for pointing that out. In that case I would prefer an early exit before https://github.com/RIOT-OS/RIOT/pull/4343/files#diff-cba0d90e0f0cd8f13f36afdda91bdb66R37 in case of `len == 0`. IMHO this is a legal situation that might happen with other buffers than `gnrc_pktbuf`. Since this is a corner case an additional unittest for the scenario you described would be great too.

authmillenon commented:
(ACK revoked until this case is fixed).

authmillenon commented:
You can squash immediately.

kb2ma commented:
Added early exit on zero-length buffer to inet_csum_slice(), added unit test for corner case, and squashed. Travis is happy.

IMHO the PR is acceptable now. However, after working with this code for a week, I do have a couple of suggestions. I would be happy to implement them if you like.

1. I think the order of the parameters for inet_csum_slice() would be improved by moving the accumulated length parameter from position 4 to position 2. It really provides context, like the checksum parameter, and belongs right behind it. Originally, I saw the function as a kind of extension to inet_csum(), and put the accumulated length parameter at the end.

2. In _calc_csum() for both UDP and ICMPv6, I would move the inet_csum() calculation for the UDP/ICMPv6 header *above* the while loop for the payload calculation. Presently, it's not clear why the 'len' local variable is initialized to hdr->size ([diff](https://github.com/RIOT-OS/RIOT/pull/4343/files#diff-a743a0dcf90d0eb32c837f1a047918c1L68)), but this initialization is essential if the header length is an odd number. (Maybe a real possibility with ICMPv6?) 

So, inet_csum_slice() would start like this:

```
uint16_t len, csum = 0;

/* process applicable UDP header bytes */
csum = inet_csum(csum, (uint8_t *)hdr->data, hdr->size);
len  = (uint16_t)hdr->size;

/* process the payload */
while (payload && payload != hdr) {
...
```

authmillenon commented:
I think both suggestions can be done as a follow-up

> 1. I think the order of the parameters for inet_csum_slice() would be improved by moving the accumulated length parameter from position 4 to position 2. It really provides context, like the checksum parameter, and belongs right behind it. Originally, I saw the function as a kind of extension to inet_csum(), and put the accumulated length parameter at the end.

I would like to look at it from the perspective what is more efficient memory-wise. Reordering parameters can lead to reassigning of registers in machine code and as a result bigger code size. Definitely something we should look into.

> In _calc_csum() for both UDP and ICMPv6, I would move the inet_csum() calculation for the UDP/ICMPv6 header above the while loop for the payload calculation. Presently, it's not clear why the 'len' local variable is initialized to hdr->size (diff), but this initialization is essential if the header length is an odd number. (Maybe a real possibility with ICMPv6?)

The length field need to be set for checksum, so I think that's why it was calculated last. But this is also something we can look into.

ACK and go for this one.

kb2ma commented:
Comments understood, and thanks for the commit and shepherding.  I see issue #3201 still is open -- might want to close that.

authmillenon commented:
Yes, thanks for the hint :-)
